[HPB-164] [FE] 建立新增文章頁

richardlinio opened the pull request:
[Issue Link](https://www.notion.so/jyu1999/FE-278cba948f1180c4a816f56f77392cb6?source=copy_link)

## Summary

引入「新增文章」功能，允許使用者透過管理後台的新表單建立和發布部落格文章。

此功能包含一個 richtext editor、立即發布或排程發布的選項，以及儲存草稿的功能。本次實作涉及建立新的文章表單、增加表單驗證、處理文章建立的 API 互動，並整合 UI 元件以提供流暢 UX。

## Major Changes

- **新增文章頁面**：在 `/admin/new-article` 新增頁面，包含用於建立文章的表單。
- **ArticleForm 元件**：建立了可複用的 `ArticleForm.vue` 元件，封裝表單邏輯，包括標題、slug、內容輸入、發佈選項欄位。
- **富文本編輯器**：整合 `TinyMceEditor.vue`，為文章內容提供所見即所得（WYSIWYG）的編輯體驗。
- **發布選項**：實作 `PublishOptions.vue`，讓使用者可以選擇立即發布、排程發布或儲存草稿。
- **表單驗證**：引入 `vee-validate` 和 `zod` 的客戶端驗證，以確保在提交前資料的完整性。
- **錯誤處理與通知**：增強了 `useArticle` composable 中的錯誤處理，並加入了 `vue-sonner`，以便在成功或失敗時提供使用者友善的 toast 通知。
- **HTML 到 Markdown 轉換**：在 `app/utils/markdown.ts` 中新增工具函式，用於在將 richtext editor 的 HTML 輸出發送到 API 之前，將其轉換為 Markdown。
- **日期/時間處理**：實作了日期/時間工具，用於處理排程發布時間，包括時區轉換。
- **設定**：
    - 在 `nuxt.config.ts` 中禁用 `/admin/**` SSR，以適應客戶端的 richtext editor 功能。
- **工具**：透過增加過濾腳本，改善了 makefile linting 和測試的輸出。

## How to Test

1. 前往新增文章頁面 `/admin/new-article`。
2. 確認表單正常顯示，包含「Title」、「Slug」、「Content」和「Publish Settings」等欄位。
<img width="3068" height="1830" alt="Screenshot 2025-10-16 at 2 30 29 PM" src="https://github.com/user-attachments/assets/aadf7b38-0c02-47a8-b103-63232ec055db" />

3. **測試表單驗證**：
    - 嘗試提交空的必填欄位（Title, Slug, Content），並確認出現驗證錯誤訊息。
    - 如果選擇「Schedule publish」，嘗試在未選擇日期和時間的情況下提交，並確認出現驗證錯誤。
<img width="3072" height="1920" alt="Screenshot 2025-10-16 at 2 31 37 PM" src="https://github.com/user-attachments/assets/928445b7-c264-47af-8fc0-75b0e249c399" />

4. **測試「Publish Now」**：
    - 填寫表單。
    - 選擇「Publish immediately」。
    - 點擊「Publish Now」。
    - 確認頁面右下角彈出「失敗通知」toast，內容為 Bad Gateway

<img width="3072" height="1920" alt="Screenshot 2025-10-16 at 2 35 39 PM" src="https://github.com/user-attachments/assets/bdbe9a34-b602-4486-9f89-5520886e39d6" />
   - 檢查網路請求 payload，確保 `published_at` 欄位為當前 UTC 時間，且欄位和 API 文件一致
<img width="1340" height="756" alt="Screenshot 2025-10-16 at 2 36 39 PM" src="https://github.com/user-attachments/assets/0f367063-7f33-498f-a099-f3ee82d90f7f" />

5. **測試「Schedule Publish」**：
    - 填寫表單。
    - 選擇「Schedule publish」並選擇一個未來的日期和時間。
    - 點擊「Schedule Publish」。
    - 確認頁面右下角彈出「失敗通知」toast，內容為 Bad Gateway
    - 檢查網路請求，確保 `published_at` 欄位為所選時間轉換後的 UTC 時間，且欄位和 API 文件一致
<img width="2716" height="1762" alt="Screenshot 2025-10-16 at 2 39 28 PM" src="https://github.com/user-attachments/assets/45789b4d-7ec9-4c96-af84-611f532ec33f" />


6. **測試「Save as Draft」**：
    - 填寫表單。
    - 選擇「Save as draft」。
    - 點擊「Save Draft」。
    - 確認頁面右下角彈出「失敗通知」toast，內容為 Bad Gateway
    - 檢查網路請求，確保 `published_at` 欄位為 `null`
<img width="2714" height="1732" alt="Screenshot 2025-10-16 at 2 40 02 PM" src="https://github.com/user-attachments/assets/29480e6c-7061-4017-900b-66669f6bc0f4" />

7. **測試內容編輯器**：
    - 使用富文本編輯器來格式化文字（粗體、斜體、列表等）。
    - 送出表單並檢查網路請求，確保內容被正確轉換為 Markdown。

## Known Issue
- [HPB-185](https://www.notion.so/jyu1999/FE-useArticle-28ecba948f11807bb159c2bf00138b0e?source=copy_link)
- [HPB-186](https://www.notion.so/jyu1999/FE-ArticleForm-vue-watchEffect-28ecba948f11804daa2fddad3523d970?source=copy_link)

richardlinio committed:
feat(routeRules): add SSR disable rule for /admin/** path
- disable server-side rendering for all /admin/** routes
- maintain existing proxy configuration for /api/** routes

richardlinio committed:
docs(plan): add detailed implementation plan for new article feature
- define goals, routes, and form fields for article creation
- specify required packages, components, and page setup
- outline button logic for saving drafts and publishing articles
- describe publish datetime input UI and behavior
- include TinyMCE editor configuration notes
- list existing resources and new requirements
- provide expected user workflow and future enhancement ideas

richardlinio committed:
feat(ui): add Alert, Label, RadioGroup, and Textarea components
- add Alert.vue, AlertDescription.vue, AlertTitle.vue with variants and styling
- add Label.vue with reactive prop delegation and styling
- add RadioGroup.vue and RadioGroupItem.vue with forwarded props and styling
- add Textarea.vue with v-model support and styling
- add index.ts files for alert, label, radio-group, and textarea components
chore(deps): upgrade reka-ui to 2.5.1 and add TinyMCE editor dependencies
- add @tinymce/tinymce-vue@^6.3.0 and tinymce@^6.8.6 to dependencies and package-lock.json

richardlinio committed:
feat(form): add form UI components with vee-validate integration
- add FormControl, FormDescription, FormItem, FormLabel, FormMessage components
- implement useFormField composable for form field context and state
- define FORM_ITEM_INJECTION_KEY for provide/inject usage
- export form components and vee-validate Form, Field, FieldArray in index.ts
- add vee-validate, @vee-validate/zod, and zod dependencies

richardlinio committed:
feat(AppSidebar): update sidebar links to admin routes
- Change 'Article list' URL to '/admin'
- Change 'New article' URL to '/admin/new-article'

feat(new-article): add new article creation page in admin panel
- Implement form with title, slug, content, publish mode (immediate or scheduled)
- Add client-side validation for required fields and scheduled date/time
- Integrate TinyMCE editor for content input with custom toolbar and plugins
- Handle save as draft and publish actions with API interaction
- Show loading states and validation error alerts
- Redirect to admin article list on successful creation

richardlinio committed:
refactor(new-article.vue): integrate vee-validate and zod for form validation and UI improvements
- replace manual form state and validation with vee-validate's useForm and zod schema
- add form components (FormField, FormItem, FormLabel, FormControl, FormMessage) for structured form layout
- update form inputs to use vee-validate bindings and validation messages
- replace manual validation error alerts with field-level error messages
- track publishMode via ref to support conditional UI rendering of scheduled datetime input
- update submit handlers to use form.handleSubmit and access validated values
- replace custom loading spinner SVGs with Loader2 component for buttons
- change form container from div to form element for semantic correctness

richardlinio committed:
chore(tinymce): downgrade TinyMCE and Vue integration versions
- Downgraded @tinymce/tinymce-vue from 6.3.0 to 4.0.7
- Downgraded tinymce from 6.8.6 to 5.10.9
- Updated import and configuration to accommodate older TinyMCE version
- Removed 'code' plugin from toolbar plugins list
- Added explicit TinyMCE skin and content CSS settings to config

richardlinio committed:
feat(new-article.vue): enhance TinyMCE editor with markdown support and paste handling
- add TinyMCE textpattern plugin with markdown-style shortcuts for inline and block formatting
- implement paste event handler to parse and convert markdown content into HTML using marked library
- add marked as a dependency for markdown parsing

richardlinio committed:
feat(article): add reusable ArticleForm component with TinyMCE editor and publish options
- create ArticleForm.vue with form fields for title, slug, content, and publish options
- integrate TinyMceEditor.vue supporting markdown paste conversion
- add PublishOptions.vue component for immediate or scheduled publishing with datetime input
- add useArticleForm composable for form validation schema and payload building
- refactor admin new-article page to use ArticleForm component and useArticleForm composable for handling form submission and validation

richardlinio committed:
refactor(article): simplify publish mode state and improve error display
- Remove redundant ref tracking for publishMode in ArticleForm.vue and rely directly on form value
- Pass validation error from ArticleForm to PublishOptions and display it below datetime input
- Clean up event handlers in PublishOptions.vue for better readability

feat(tinymce): improve paste event typing and rename editor config variable
- Use TinyMCEEditor type for editor in setup function
- Rename tinymceConfig to editorConfig for clarity

feat(new-article): add notification system for create and publish actions
- Replace alert calls with UI Alert component notifications including success and error states
- Add auto-hide timeout and delayed navigation after notifications
- Fix event handler naming consistency (handleSaveAsDraft → handleSaveDraft)

docs(plan): expand future enhancement plan for article features and UX
- Add detailed items for article management, form experience, datetime handling, notifications, editor features, and code quality improvements

richardlinio committed:
docs(TODO.md): add detailed refactor task list for HPB Blog Frontend article editor
- define project goals and core requirements for markdown-based content editing
- outline phased implementation plan including markdown utils, editor refactor, architecture reorganization, notification abstraction, and testing
- specify installation steps and utility functions for markdown/html conversion
- propose TinyMCE plugin adjustments and paste event handling improvements
- introduce new type definitions and validation schema modularization
- refactor payload builders to convert editor HTML content to markdown before submit
- extract notification composable for unified usage
- plan comprehensive unit, component, and e2e tests with Playwright and Vitest
- include acceptance criteria and future optimization notes

richardlinio committed:
feat(package): add turndown and its type definitions
- add turndown dependency version 7.2.1
- add @types/turndown dependency version 5.0.5
- update package-lock.json to include turndown and related packages

richardlinio committed:
feat(utils/markdown): add HTML to Markdown and Markdown to HTML conversion utilities
- Introduce `htmlToMarkdown` function using TurndownService with GFM plugin and custom rules for headings and strikethrough
- Introduce `markdownToHtml` function using marked with GFM enabled and breaks disabled
- Add comprehensive unit tests covering headings, formatting, lists, links, images, code blocks, blockquotes, horizontal rules, tables, GFM features, edge cases, and round-trip conversions
- Add `turndown-plugin-gfm` dependency
- Fix vitest config to include nested test files in unit and nuxt test folders

richardlinio committed:
docs(types): add type declarations for turndown-plugin-gfm
- declare module for 'turndown-plugin-gfm'
- define GfmPluginOptions interface with optional highlightCode
- export gfm, tables, strikethrough, taskListItems, and autolink plugins types

richardlinio committed:
perf(Makefile): enhance test-llm output filtering

- extended grep filter patterns to exclude more verbose Vue and Nuxt debug logs during test runs
- reduced noise in test output for clearer visibility of relevant test results

richardlinio committed:
feat(utils/markdown): add detectMarkdown function for Markdown syntax detection
- Implement detectMarkdown to identify common Markdown patterns (headers, bold, italic, links, lists, blockquotes, code blocks)
- Return false for empty, whitespace-only, or non-string inputs
- Add comprehensive unit tests covering headers, formatting, links, lists, blockquotes, code blocks, combined markdown, negative cases, edge cases, and false positive prevention

richardlinio committed:
refactor(TinyMceEditor.vue): replace inline markdown detection and parsing with utilities
- removed manual markdown syntax checks and replaced with detectMarkdown utility
- replaced marked.parse usage with markdownToHtml utility for markdown to HTML conversion

richardlinio committed:
refactor(tinymce): simplify editor configuration and remove markdown paste handling
- Removed multiple unused TinyMCE plugins to streamline the editor setup
- Disabled menubar and updated toolbar buttons for a cleaner UI
- Removed custom paste event handler that converted markdown to HTML on paste

refactor(markdown): remove detectMarkdown function and related tests
- Deleted detectMarkdown utility function and all associated unit tests from markdown utilities

richardlinio committed:
style(Makefile): improve test-llm output filtering and cleanup
- Add cat -s to test-llm target to squeeze multiple blank lines into one for cleaner output

richardlinio committed:
refactor(form): unify ArticleFormValues type and rename variables for clarity
- Extract ArticleFormValues type to a dedicated file (app/types/form.ts)
- Rename variables from generic 'values' to 'articleFormValues' in handlers and emits
- Rename 'handleSaveAsDraft' to 'handleSaveDraft' for consistency
- Rename form variable from 'form' to 'articleForm' in ArticleForm.vue
- Update imports to use the new ArticleFormValues type path across components and composables

richardlinio committed:
refactor(article): separate validation schema and payload builder from useArticleForm composable
- rename and move article validation schema to `useArticleValidation.ts`
- rename and move payload builder to `useArticlePayload.ts`
- remove deprecated `useArticleForm.ts` composable
- update imports in ArticleForm.vue and new-article.vue accordingly
- enhance buildNewArticlePayload to convert HTML content to Markdown
test(useArticlePayload): add unit tests for buildNewArticlePayload
- cover HTML to Markdown conversion correctness
- verify published_at handling for draft, immediate publish, and scheduled publish
- test edge cases with empty and plain text content
docs(TODO.md): update composable filenames and clarify payload builder refactor details

richardlinio committed:
refactor(useArticlePayload): unify article action handling and add scheduled publish support
- Replace boolean isDraft parameter with explicit ArticleAction type for clarity
- Introduce getPublishedAt helper to determine published_at timestamp based on action
- Update new-article.vue to use ArticleAction for save-draft, publish-immediate, and publish-scheduled actions
- Extend ArticleAction type to include 'save-draft', 'publish-immediate', and 'publish-scheduled'
- Add tests for scheduled publish mode and missing scheduledDateTime error handling

richardlinio committed:
refactor(article): unify publishSetting handling and add save-draft option

- Rename publishMode to publishSetting with values: 'publish-immediate', 'publish-scheduled', and 'save-draft'
- Update ArticleForm.vue to handle new publishSetting including computed properties and button states
- Modify PublishOptions.vue to support new publishSetting radio options and UI
- Adjust useArticlePayload to derive publish action from publishSetting only
- Update validation schema to validate publishSetting with new values and enforce scheduledDateTime requirement accordingly
- Refactor new-article page to use updated buildNewArticlePayload signature and handle save-draft action properly
- Modify form types to use publishSetting instead of publishMode
- Update unit tests for useArticlePayload to reflect new publishSetting usage and save-draft scenario

richardlinio committed:
refactor(article): unify submit handling and rename variables for clarity
- Replace separate saveDraft and publish emits with a single submit event in ArticleForm.vue
- Rename isScheduledMode to isScheduledPublish and related variables for consistency
- Simplify button actions in ArticleForm.vue to use a single submit handler
- Consolidate success message logic and submit handler in new-article.vue
- Update PublishOptions.vue to use isScheduledPublish for clarity

richardlinio committed:
feat(alert): add composable for alert state management

- Create useAlert composable to manage alert visibility, variant, title, and message
- Implement auto-hide functionality with customizable duration and proper timeout cleanup
- Define alert constants for auto-hide duration and navigation delay after success
- Add TypeScript types for alert state and show options
- Refactor new-article page to use useAlert composable instead of local notification state
- Replace notification UI logic with alert composable reactive state
- Add comprehensive unit tests for useAlert composable covering show, hide, auto-hide timing, and timeout clearing

richardlinio committed:
feat(alert): add dismissible alerts with manual close control
- Add dismissible prop and close button to Alert.vue component
- Emit close event to allow manual hiding of alert
- Enhance useAlert composable to support null duration for persistent alerts
- Prevent auto-hide timeout when duration is null
- Add hideAlert method to manually hide alerts
- Update new-article page to clear alerts before showing new ones and prevent auto-hide on error alerts
- Add unit tests for alert duration null behavior, immediate hide, and manual close

richardlinio committed:
refactor(alert): standardize alert auto-hide duration handling

- use Infinity instead of null to indicate never auto-hide in alert duration
- update type definitions to reflect Infinity as never auto-hide
- modify useAlert composable to check for finite duration instead of null
- update test cases to use Infinity for never auto-hide scenarios
- add default variant and class props to Alert component
- change new-article page error alert duration to Infinity to prevent auto-hide

richardlinio committed:
fix(ui/alert): add border to destructive variant styling
- include 'border-destructive' class in destructive alert variant for proper border display

richardlinio committed:
style(alert): clean up comments and improve consistency
- Remove outdated multiline comments in useAlert.ts about auto-hide durations
- Simplify and convert block comments to single-line in alert.ts
- Remove commented note in new-article.vue about API error auto-hide behavior
- Clarify duration comment in alert.ts type definition to reference useAlert() setting

richardlinio committed:
refactor(useArticle): remove error state handling from composable and tests
- Removed error state and related error handling logic in useArticle composable
- Updated tests to remove assertions and usage of error state
- Simplified new-article.vue to handle errors via getErrorMessage utility instead of composable error state
- Added error helper utilities (ApiError class, isApiError, getErrorMessage) with unit tests

richardlinio committed:
refactor(test): move useAlert test file to nuxt folder
- renamed useAlert.test.ts from test/unit/composables to test/nuxt directory

richardlinio committed:
refactor(useArticleValidation): reorder string trim and min validations
- apply trim() before min() to ensure whitespace is trimmed prior to length check for title, slug, and content fields

richardlinio committed:
style(PublishOptions.vue): remove extra spaces in label text
- Trimmed whitespace in "Save as draft" label for cleaner UI display

richardlinio committed:
style(PublishOptions.vue): add type annotation and adjust spacing
- add explicit string type annotation to event handler parameter
- add spaces inside label text for consistency

richardlinio committed:
chore(docs): remove outdated project planning documents

- deleted TODO.md containing detailed refactoring tasks, priorities, and implementation plans for the blog frontend
- removed plan.md outlining the new article feature implementation steps, technical details, and future enhancements

richardlinio committed:
docs(FIX): add comprehensive code review fix list and suggestions

- Add detailed P0 critical fixes including useAlert memory leak fix with cleanup on unmount
- Restore error handling in useArticle composable with error state and try-catch blocks
- Correct scheduled publish logic in useArticlePayload to keep status as draft before scheduled time
- Suggest fixes for P1 important issues: prevent duplicate submits in ArticleForm.vue, sync props update to form, add type validation in PublishOptions.vue, and delay navigation after success in new-article.vue
- Provide P2 optional improvements for TinyMceEditor CSS consistency, timezone handling in validation, markdown code block processing, error format extension, Makefile test output filtering, and add E2E test recommendations
- Include security checklist for XSS, CSRF, rate limiting, and permission validation
- Outline phased fix priority and checklist for completion
- Reference relevant external documentation and best practices

richardlinio committed:
refactor(useAlert): add cleanupTimeout function and unregister timeout on unmount
- Extract timeout clearing logic to cleanupTimeout helper function
- Register onUnmounted lifecycle hook to clear timeout when component unmounts
- Update showAlert and hideAlert to use cleanupTimeout for timeout management
test(useAlert): add test to verify timeout cleanup on component unmount
- Add test component to mount and trigger alert with timeout
- Unmount component and assert clearTimeout is called
- Ensure no errors occur after timeout would have fired following unmount

richardlinio committed:
feat(ui/sonner): add Sonner toaster component integration

- add Sonner.vue component wrapping vue-sonner Toaster with custom styles
- export Toaster component from sonner/index.ts
- add vue-sonner dependency to package.json and package-lock.json

richardlinio committed:
feat(app): add toast notifications for article API errors

- Integrate vue-sonner Toaster component globally in app.vue
- Add error state and toast error messages to useArticle composable for all API calls
- Remove redundant error alert handling in new-article page since errors are handled by toast notifications

richardlinio committed:
refactor(useArticle): extract common error handling logic

- Add handleError function to centralize API error processing and toast notification
- Replace repeated error handling code in list, get, create, update, and delete functions with handleError calls

richardlinio committed:
refactor(new-article.vue): simplify article creation flow and replace alert with toast
- remove useAlert composable and alert component usage
- replace alert with vue-sonner toast for success message
- streamline handleSubmit by removing try-catch and alert state management
- keep redirection to /admin after success with delay

richardlinio committed:
test(nuxt): add comprehensive tests for new-article page and useArticle composable
- add extensive unit tests for NewArticlePage including submission handling, toasts, navigation, and loading state
- test error handling and error state management in useArticle composable for all CRUD operations
- mock vue-sonner toast methods in tests to verify success and error notifications
- add utility function createApiErrorInstance for generating real ApiError instances in tests
- improve api-mock-data with real ApiError instance creation helper for composable error state testing

richardlinio committed:
refactor(alert): remove alert UI components and composable logic
- delete Alert.vue, AlertDescription.vue, AlertTitle.vue components
- remove alert variant styles and exports from index.ts
- remove useAlert composable and its associated logic and lifecycle management
- delete all related tests for useAlert composable

richardlinio committed:
feat(ArticleForm): prevent duplicate submissions and improve loading state UX
- Added check in submit handler to prevent submission during loading state
- Changed form to use @submit.prevent and wrapped inputs in a disabled fieldset when loading
- Updated submit button to type="submit" and disabled during loading
- Added visual feedback (opacity and pointer-events) to fieldset when loading
test(ArticleForm): add comprehensive tests for submission and loading behavior
- Added tests to verify submit event emission respects loading state
- Verified UI updates correctly when loading prop changes
- Tested form structure with fieldset wrapping inputs and submit button outside
- Mocked dependent components and vee-validate for isolated testing

richardlinio committed:
fix(useAlert): remove deprecated fix instructions and outdated test details from FIX.md
- deleted detailed fix plan and test related to memory leak issue in useAlert.ts to clean up documentation

fix(useArticle): remove deprecated error handling discussion and partial fix plans from FIX.md
- removed extensive explanation and code snippets about restoring error state in useArticle composable and test updates

fix(useArticlePayload): remove scheduled publish status fix details from FIX.md
- deleted description and code samples correcting publish status for scheduled articles and related test case

fix(ArticleForm.vue): remove repetitive submit prevention fix notes from FIX.md
- removed fix suggestions and code snippets about preventing duplicate submission on loading state

docs(FIX.md): clean up large removed sections and outdated fix discussions
- drastically reduced size of FIX.md by removing multiple large outdated fix proposals and test details related to useAlert, useArticle, useArticlePayload, and ArticleForm.vue

richardlinio committed:
feat(ArticleForm): add reactive reset of form on initialValues prop update
- initialize form with empty default values instead of props initially
- watch initialValues prop and reset form when it changes if form is not dirty and initialValues is non-empty
- update tests to mock resetForm and meta.dirty for verifying reactive reset behavior

richardlinio committed:
refactor(article): validate and handle publishSetting update safely
- add ARTICLE_ACTIONS constant for valid article actions
- implement isValidArticleAction type guard to check action validity
- replace inline emit with handlePublishSettingUpdate method that validates input before emitting
- update ArticleAction type to derive from ARTICLE_ACTIONS array

richardlinio committed:
refactor(new-article.vue): simplify navigation handling after article creation
- Remove alert constants and related delayed navigation logic
- Replace useRouter and setTimeout with direct await navigateTo call
- Remove related tests for delayed navigation and fake timers
- Add requestAnimationFrame polyfill in test setup to support test environment

richardlinio committed:
feat(datetime): add datetime validation and conversion utilities
- introduce VALID_YEAR_MIN and VALID_YEAR_MAX constants for valid year range
- add convertLocalToUTC to convert local datetime string to UTC ISO string with validation
- add isValidDateTime function to validate datetime string and enforce year range
- update buildNewArticlePayload to use convertLocalToUTC for scheduled publish datetime
- enhance articleFormSchema validation to check scheduledDateTime presence and validity with custom messages
- add error handling in new-article.vue submit handler with toast error notification
- add comprehensive unit tests for datetime utils, article payload, and validation composables
- add date-fns dependency for datetime parsing and validation

richardlinio committed:
chore(scripts): add output filter scripts for lint, typecheck, and test commands
- created filter-eslint-output.sh to clean eslint output from Nuxt/Nitro debug messages
- created filter-typecheck-output.sh to clean nuxt typecheck output including performance messages
- created filter-test-output.sh to clean vitest run output from various noisy logs and Vue internals
- updated Makefile to use these filter scripts for lint-llm and test-llm targets

richardlinio committed:
docs(FIX): remove obsolete code review fix list document
- delete the entire FIX.md file containing detailed code review issues and suggestions

richardlinio committed:
refactor(types): remove alert type definitions
- deleted AlertState interface
- deleted ShowAlertOptions interface

richardlinio committed:
docs(FIX): add detailed issue fix proposals and improvement suggestions

- add high priority fix for duplicate error toast notifications with two alternative solutions
- propose medium priority fixes including replacing watchEffect with watch in ArticleForm, safer type guard use in useArticle, and dynamic loading or CDN use for TinyMCE to reduce bundle size
- provide low priority suggestions for better error handling in PublishOptions, avoiding redundant fieldset styles in ArticleForm, and handling unused getUserTimezone function

richardlinio committed:
refactor(utils/datetime): remove unused getUserTimezone function and related tests
- deleted getUserTimezone function from datetime utilities
- removed getUserTimezone tests from datetime unit tests

richardlinio committed:
style(article/ArticleForm.vue): simplify loading state styles on fieldset
- remove 'pointer-events-none' class binding while keeping opacity effect
test(nuxt/ArticleForm.test.ts): update tests to reflect removal of pointer-events-none class
- remove assertions related to 'pointer-events-none' class presence and absence

richardlinio committed:
refactor(docs): remove FIX.md issue tracking document
- delete detailed fix and improvement notes related to critical, medium, and low priority issues
- remove explanations and code samples addressing error notification, watchEffect usage, type safety, TinyMCE import optimization, error handling, style duplication, and unused functions

richardlinio committed:
fix(article): use toast for invalid action error in PublishOptions.vue
- Replace console.error with toast.error for invalid article action

refactor(useArticle): handle errors without throwing exceptions
- Change handleError to return void instead of never and remove throw
- Update async article functions to return undefined on error instead of throwing

fix(new-article): handle undefined create response gracefully
- Remove try-catch in handleSubmit and check for newArticle before showing success toast and navigation

test(new-article): improve create mock and tests
- Add createMockArticle helper for consistent mock article data
- Update create mockResolvedValue to return mock article instead of undefined in new-article tests

test(useArticle): adapt tests for error handling changes
- Replace expect(...).rejects.toEqual(...) with result checks for undefined
- Verify toast.error calls and loading states without expecting thrown exceptions
- Update remove tests to await calls and assert error handling internally without throwing

richardlinio committed:
test(nuxt): add requestAnimationFrame polyfill for Nuxt environment
- add setup file to polyfill requestAnimationFrame and cancelAnimationFrame in test/nuxt/setup.ts
- include setup file in vitest config for nuxt environment tests

KR0924 commented on app/constants/datetime.ts:
```diff
@@ -0,0 +1,2 @@
+export const VALID_YEAR_MIN = 1970
+export const VALID_YEAR_MAX = 2100
```
為什麼要給他最小1970這數字，看起來是用在設定public的時間

richardlinio replied:
我是覺得沒有最小值會不太合理，至少也不能變成負的。

以部落格而言，就算有人想要搬自己舊的文章，總不會比 1970 年代個人電腦發明還早，所以我設這個時間。

如果是有需要存紙本發表過的文章的話，可以開得更早。

總體來說有個合理的數字就好

KR0924 replied:
MVP階段我會認為這是個合理的東西。
但以正常流程來想讓使用者可以設定發布時間是今天以前有點怪怪的

richardlinio replied:
目前市面上的其他幾個平台：
1. wordpress 可以改成過去時間
2. medium 應該只能刪掉或編輯
3. substack 可以 unpublish，應該是變回草稿

有些出版條件會限定首次出版，因此我認為發佈時間不可晚於現在是合理的，否則該出版平台將不被信任
至於 1.0 部落格舊文章搬家，發佈時間在過去，但能方便搬家這應該是另一個功能了，有需要的話再另開

結論：同意排程發佈時間必須晚於現在時間，否則拋出錯誤訊息
以上已更改於 commit 9b4fccfc6db5946127d5d577fc7fb8c354341b98

KR0924 commented on app/components/ui/form/FormDescription.vue:
```diff
@@ -0,0 +1,21 @@
+<script lang="ts" setup>
+import type { HTMLAttributes } from 'vue'
+import { cn } from '@/lib/utils'
+import { useFormField } from './useFormField'
+
+const props = defineProps<{
+  class?: HTMLAttributes['class']
+}>()
+
+const { formDescriptionId } = useFormField()
+</script>
+
+<template>
+  <p
+    :id="formDescriptionId"
+    data-slot="form-description"
+    :class="cn('text-muted-foreground text-sm', props.class)"
+  >
+    <slot />
+  </p>
+</template>
```
這檔案未來的作用是什麼？
我目前好像沒看到引用這隻檔案

richardlinio replied:
這個檔案是 UI library 的一部分，即便這次我沒有使用，Shadcn Form 元件邏輯層的 `index.ts`, `useFormField.ts` 等檔案都有 import 並使用這個檔案的資料，如果刪除它那 ESLint 會馬上報錯，需要去動它元件邏輯層的 code。

基於 Nuxt 的 tree shaking 機制，多餘的 code 並不會被編譯到 runtime，應不致於對效能產生太大影響。

綜合起來，我會建議保留，維持 UI library 檔案的完整性。Form Description 元件我沒用到，最多就是 ui/ 資料夾中有一些尚未使用的 code，但要砍的話要改它元件的邏輯層，我認為是不必要的風險

KR0924 commented on app/components/article/ArticleForm.vue:
```diff
@@ -0,0 +1,146 @@
+              @update:scheduled-date-time="
+                (val) => articleForm.setFieldValue('scheduledDateTime', val)
+              "
+            />
```
這段不考慮放進 app/components/article/PublishOptions.vue 中的原因是什麼？

richardlinio replied:
整個放進去會有的問題是，`PublishOptions` 需要 `ArticleForm` 裡面的 `scheduledDateTime`, `isSchedulesPublish` 這兩個資料，這兩者因為影響到 submit button 的字樣，一定會在 `ArticleForm` 這一層，而非 `PublishOptions` 內部的變數。目前的作法就是把這兩個資料再傳入到 `PublishOptions` 元件中，我想這是合理的

不過也感謝你的提示，我想到 `scheduledDateTime` 目前的管理方式還有優化空間。目前是手動管理欄位值，但因為他是 `PublishOptions` 中的一個欄位，我可以把它當成一個表單中的表單，以此重複使用 vee validate 套件

因此我決定用具名表單的方式處理，這樣在 `PublishOptions` 中就不用額外的 v-if 處理欄位錯誤資訊，而是可以使用 vee validate 的 `FormMessage`。這樣你後面的另一個 comment 也就一起處理掉了。

以上已修改於 commit d2bcf03cb754f8541b9b2fe80cc1de0d6af0e341

KR0924 commented on app/components/article/TinyMceEditor.vue:
```diff
@@ -0,0 +1,76 @@
+
+// TinyMCE configuration
+const editorConfig = {
+  height: 500,
```
這邊我的理解是任何裝置進來都會是固定500的高度嗎？

richardlinio replied:
你的理解正確。一開始是我覺得桌面版的預設起始高度太低，但這樣設會讓手機版太高，而桌面版元件的右下角也有可以拖曳拉高的地方。

綜合來說，我決定刪掉這行，先保持預設值，結語使用起來如果覺得確實預設高度太低，我們再針對桌面版調整高度。

已處理於 commit edcde6c9e3e7627adbbba19a717962406c549ad5

KR0924 commented on app/components/article/PublishOptions.vue:
```diff
@@ -0,0 +1,76 @@
+        />
+        <p v-if="error" class="text-destructive text-sm font-medium">
+          {{ error }}
+        </p>
```
這邊的錯誤會跟 app/components/ui/form/FormMessage.vue 裡面的不一樣嗎？

richardlinio replied:
已於[此 comment](https://github.com/Goodidea-backend-camp/hpb-blog-frontend/pull/8/files#r2443905308) 回覆，這段 code 已在 commit d2bcf03cb754f8541b9b2fe80cc1de0d6af0e341 修改了

KR0924 replied:
[此 comment](https://github.com/Goodidea-backend-camp/hpb-blog-frontend/pull/8/files#r2443905308)我好像看不到

linporu replied:
因為你按 resolve 了嗎？就是之前你問為什麼不整段塞到 publishOptions 那邊

richardlinio replied:
我先貼過來：

整個放進去會有的問題是，PublishOptions 需要 ArticleForm 裡面的 scheduledDateTime, isSchedulesPublish 這兩個資料，這兩者因為影響到 submit button 的字樣，一定會在 ArticleForm 這一層，而非 PublishOptions 內部的變數。目前的作法就是把這兩個資料再傳入到 PublishOptions 元件中，我想這是合理的

不過也感謝你的提示，我想到 scheduledDateTime 目前的管理方式還有優化空間。目前是手動管理欄位值，但因為他是 PublishOptions 中的一個欄位，我可以把它當成一個表單中的表單，以此重複使用 vee validate 套件

因此我決定用具名表單的方式處理，這樣在 PublishOptions 中就不用額外的 v-if 處理欄位錯誤資訊，而是可以使用 vee validate 的 FormMessage。這樣你後面的另一個 comment 也就一起處理掉了。

richardlinio committed:
refactor(article): move scheduledDateTime input to slot in PublishOptions component
- Remove scheduledDateTime prop and related error handling from PublishOptions
- Replace internal datetime input in PublishOptions with a named slot "scheduled-date-time"
- Update ArticleForm to provide scheduledDateTime input via the slot with form validation
- Adjust test mocks to match updated PublishOptions props and slot usage

richardlinio committed:
style(TinyMceEditor.vue): remove fixed editor height from config
- deleted height property from TinyMCE editor configuration to allow flexible height

richardlinio committed:
refactor(datetime): rename and improve publish datetime validation logic
- Rename isValidDateTime to isValidPublishDateTime to better reflect validation purpose
- Update validation to require scheduled publish datetime to be strictly in the future and not beyond max year
- Adjust article form validation to use updated isValidPublishDateTime with new error message
- Enhance unit tests for isValidPublishDateTime to cover future date requirements, invalid past dates, boundary conditions, leap years, and various datetime formats
- Remove old minimum year boundary validation and related tests no longer applicable
